migration-from-4x.md

```diff
@@ -0,0 +1,328 @@
+  }
+});
+```
+> Note: The callback function given to `onSetup` should take a `buttonApi` argument which is an object that contains some helper functions.
+
+Each type of toolbar button has a different API. In this example, the button's API contains `isDisabled: () => boolean` and `setDisabled: (state: boolean) => void`.
+
+Also, `onpostrender` was only ever called once, when the editor is initialized. `onSetup` is called whenever the component is rendered, which may be multiple times. This means we may need a `onTearDown` handler. Rather than have to define this as a separate configuration option, `onSetup` can return a function to be used on tear down.
+
+This is particularly important if `onSetup` listens to any events using `editor.on(eventName, callback)`. Unless the event was `'init'`, `onSetup` should return `(buttonApi) => ed.off(eventName, callback)`. The **teardown** function will automatically be called by the editor when necessary.
+
+If some functionality should only run once, when the editor is initialized, it should be passed to `editor.on('init', callback)` as the callback function so it's only run when the editor is initialized.
+
+### Custom Menu Items
+
+##### New methods:
+
+| **New Method** | **Description** |
+| ----------- | -------------- |
+| editor.ui.registry.addMenuItem() | Adds a custom basic menu item. |
+| editor.ui.registry.addToggleMenuItem() | Adds a custom toggle menu item. |
+
+Docs coming soon!
+
+### Custom Dialogs
+
+Dialogs no longer have `height` or `width` settings. The dialog component now uses CSS3 and a predefined `small`, `medium`, and `large` template to specify the dimensions.
+
+### Custom Sidebars
+
+Docs coming soon!
+
+### Custom Context Toolbars
+
+The Context Toolbar takes its buttons the registry of toolbar buttons added using `addButton`, `addToggleButton`, `addSplitButton` or `addMenuButton`. The method for creating custom context toolbars has also been moved from `editor.addContextToolbar()` to `editor.ui.registry.addContextToolbar()`.
+
+For more information on Context Toolbars, see [docs]({{site.baseurl}}/components/contexttoolbar).
+
+#### Changes between TinyMCE 4.x and TinyMCE 5.0:
+
+* Buttons go before and after the input in TinyMCE 4.x.
+* The `Ctrl+K` shortcut does nothing until the context toolbar is visible in TinyMCE 4.x.
+* In TinyMCE 5.0, the pop animates to its new width.
+* In TinyMCE 4.x., it is a URL input, so you get a popup and a browse button.
+
+### Custom Context Menus
+
+The default context menu is no longer a plugin, it is part of the core and always enabled. Where TinyMCE 4.x only supported adding registered menu items, the new context menu also allows plugins to register "sections" of the context menu. These sections are dynamic and may show or hide depending on the cursor position when the context menu is opened. For example, the default context menu config is now `'link image editimage table spelling'` which are all plugin section references, not menu items.
```

Comment:
I think the first sentence needs to be expanded, or another sentence needs to be added. It doesn't flow to the next sentence.

Reply:
Reference link added.